crypto/cmp/cmp_ctx.c

```diff
@@ -0,0 +1,1276 @@
+{
+    int i;
+
+    if (ctx == NULL)
+        return 0;
+    if (fail_info == NULL)
+        return 1;
+
+    ctx->failInfoCode = 0;
+    for (i = 0; i <= OSSL_CMP_PKIFAILUREINFO_MAX; i++)
+        if (ASN1_BIT_STRING_get_bit(fail_info, i))
+            ctx->failInfoCode |= (1 << i);
+
+    return 1;
+}
+
+/*
+ * Get the failinfo error code bits in OSSL_CMP_CTX
+ * returns bit string as integer on success, -1 on error
+ */
+int OSSL_CMP_CTX_failInfoCode_get(OSSL_CMP_CTX *ctx)
+{
+    if (ctx == NULL)
+        return -1;
+    return ctx->failInfoCode;
+}
+
+#if 0
+/*
+ * pushes a given 0-terminated character string to ctx->freeText
+ * this is intended for human consumption
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_push_freeText(OSSL_CMP_CTX *ctx, const char *text)
+{
+    if (ctx == NULL)
+        goto err;
+
+    ctx->freeText = CMP_PKIFREETEXT_push_str(ctx->freeText, text);
+    return ctx->freeText != NULL;
+
+ err:
+    CMPerr(CMP_F_OSSL_CMP_CTX_PUSH_FREETEXT, CMP_R_NULL_ARGUMENT);
+    return 0;
+}
+#endif
+
+/*
+ * sets a BOOLEAN or INT option of the context to the "val" arg
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_set_option(OSSL_CMP_CTX *ctx, OSSL_cmp_option opt, int val) {
+    if (ctx == NULL)
+        goto err;
+    switch (opt) {
+    case OSSL_CMP_OPT_IMPLICITCONFIRM:
+        ctx->implicitConfirm = val;
+        break;
+    /* to cope with broken server ignoring implicit confirmation */
+    case OSSL_CMP_OPT_DISABLECONFIRM:
+        ctx->disableConfirm = val;
+        break;
+    case OSSL_CMP_OPT_UNPROTECTED_SEND:
+        ctx->unprotectedSend = val;
+        break;
+    case OSSL_CMP_OPT_UNPROTECTED_ERRORS:
+        ctx->unprotectedErrors = val;
+        break;
+    case OSSL_CMP_OPT_VALIDITYDAYS:
+        ctx->days = val;
+        break;
+    case OSSL_CMP_OPT_SUBJECTALTNAME_NODEFAULT:
+        ctx->SubjectAltName_nodefault = val;
+        break;
+    case OSSL_CMP_OPT_SUBJECTALTNAME_CRITICAL:
+        ctx->setSubjectAltNameCritical = val;
+        break;
+    case OSSL_CMP_OPT_POLICIES_CRITICAL:
+        ctx->setPoliciesCritical = val;
+        break;
+    case OSSL_CMP_OPT_IGNORE_KEYUSAGE:
+        ctx->ignore_keyusage = val;
+        break;
+    case OSSL_CMP_OPT_POPOMETHOD:
+        ctx->popoMethod = val;
+        break;
+    case OSSL_CMP_OPT_DIGEST_ALGNID:
+        ctx->digest = val;
+        break;
+    case OSSL_CMP_OPT_MSGTIMEOUT:
+        ctx->msgtimeout = val;
+        break;
+    case OSSL_CMP_OPT_TOTALTIMEOUT:
+        ctx->totaltimeout = val;
+        break;
+    case OSSL_CMP_OPT_PERMIT_TA_IN_EXTRACERTS_FOR_IR:
+        ctx->permitTAInExtraCertsForIR = val;
+        break;
+    case OSSL_CMP_OPT_REVOCATION_REASON:
+        ctx->revocationReason = val;
+        break;
+    default:
+        goto err;
+    }
+
+    return 1;
+  err:
+    return 0;
+}
+
+/*
+ * Function used for outputting error/warn/debug messages depending on
```

Comment:
I have not been aware of the (rather new) trace API so far.
While that API clearly has a similar purpose, the logging facility used here for CMP has several advantages:

- it supports the concept of severity levels, e.g., debug, info, warning, error, ...
- it adds source information (file name, function name, line number) automatically
- it is simpler to use, in particular regarding string formatting and variadic args

I've now adapted this to using the trace API underneath, while keeping its original flexibility.
Doing so, I've added a proposal for extending the trace categories defined in `trace.h` by
```
# define OSSL_TRACE_CATEGORY_FATAL              13
# define OSSL_TRACE_CATEGORY_ERR                14
# define OSSL_TRACE_CATEGORY_WARN               15
# define OSSL_TRACE_CATEGORY_INFO               16
# define OSSL_TRACE_CATEGORY_DEBUG              17
```
What do you think?

Reply:
@mattcaswell, could you please already have a look at my commit https://github.com/mpeylo/cmpossl/pull/178/commits/d897f171c3e2f56e37174a9231521eaa183388e7 and comment on it? 
Is this use of the trace API and my provisional extension I did to its categories fine by you?

Reply:
I'm away at the ICMC conference at the moment. Will try an take a look at your latest updates next week.

Reply:
Thanks for letting us know. Have a good (remaining) conference!

Reply:
I'm feeling quite opposed to the idea of introducing a CMP specific logging capability (even one based on the trace API underneath). I really would not expect to see any CMP logging functions in the public API. If the trace API is not sufficient then we should rather change that so that it *is* sufficient. I'd also think that such changes to the trace API should be pulled out into a separate PR so that they can be more easily seen by the broader community.

Reply:
All right, will do that.
In this way also others will be able to benefit from the enhanced trace API.

Reply:
In preparation of a separate PR for the trace API improvements and CMP chunk 4 (which I plan for later this week) I've just moved the enhancements of the trace API (namely, a severity/verbosity level and automatic output of function name, file name, line number, and severity level) to `trace.h` and `trace.c` and did some small further improvements.
@mattcaswell, what do you think about the new trace/logging code?

BTW, it turns out that trace facility of OpenSSL is disabled globally by default and needs to be enabled explicitly via the `enable-trace` configuration option. This is not adequate when using the trace API for error and warning output.
What I propose instead in the future is to enable it by default, by using `OSSL_TRACE_LEVEL_INFO` as the default level of verbosity, which disables detailed (trace-level and debugging) output while enabling alert/fatal, error, warning, and info output.
Using my new code, the level of verbosity can be adapted per category.